### PR TITLE
Add struct pulling to ghidra

### PR DESCRIPTION
**Pull request by Flipout50**

### CODE REVIEW

File: binsync/decompilers/ghidra/controller.py [hunk 1]
```diff
@@ -130,9 +130,23 @@ def fill_function_header(self, func_addr, user=None, artifact=None, decompilatio
...
     @ghidra_transaction
     def fill_struct(self, struct_name, header=True, members=True, artifact=None, **kwargs):
-        struct: Struct = artifact;
-        # TODO: fill structs
-
+        struct: Struct = artifact
+        ghidra_struct = self._get_struct_by_name(struct_name)
+        data_manager = self.ghidra.currentProgram.getDataTypeManager()
+        # TODO: Also update existing struct instead of making a new one
+        self.ghidra_bridge.remote_exec(
```
Comment (mahaloz) [inline binsync/decompilers/ghidra/controller.py:137]: You don't need to do `remote_exec` like this if the code is not slow. In this case, it's not even a loop, so I think it should be find to just write that code here.

**Review by mahaloz — CHANGES_REQUESTED** [review]

File: binsync/decompilers/ghidra/controller.py (hunk 1 above)
Comment (Flipout50) [reply binsync/decompilers/ghidra/controller.py:137]: Ok, good to know

File: binsync/decompilers/ghidra/controller.py [hunk 2]
```diff
@@ -130,8 +130,40 @@ def fill_function_header(self, func_addr, user=None, artifact=None, decompilatio
...
+            for offset in struct.members:
+                member = struct.members[offset]
+                ghidra_struct.add(byteType.dataType, 1, member.name, "")
+                ghidra_struct.growStructure(member.size-1)
+                for dtc in ghidra_struct.getComponents():
+                    if dtc.getFieldName() == member.name:
+                        if member.type:
+                            gtype = self.typestr_to_gtype(member.type)
+                        else:
+                            gtype = self.typestr_to_gtype('undefined'+str(member.size))
```
Comment (mahaloz) [inline binsync/decompilers/ghidra/controller.py:152]: The naming style we are going for is `field_0`, where `0` is the offset in the struct in hex (not the size). Fix this

File: binsync/decompilers/ghidra/controller.py [hunk 3]
```diff
@@ -130,8 +130,40 @@ def fill_function_header(self, func_addr, user=None, artifact=None, decompilatio
...
+            structType = self.ghidra.import_module_object("ghidra.program.model.data", "StructureDataType")
+            byteType = self.ghidra.import_module_object("ghidra.program.model.data", "ByteDataType")
+            ghidra_struct = structType(struct_name, 0)
+            for offset in struct.members:
+                member = struct.members[offset]
+                ghidra_struct.add(byteType.dataType, 1, member.name, "")
+                ghidra_struct.growStructure(member.size-1)
+                for dtc in ghidra_struct.getComponents():
+                    if dtc.getFieldName() == member.name:
+                        if member.type:
```
Comment (mahaloz) [inline binsync/decompilers/ghidra/controller.py:149]: Any time you find yourself doing a single assignment or call based on 1 or 2 conditions, you should use inline conditional checks to make it Pythonic: 
```suggestion
                        gtype = self.typestr_to_gtype(member.type if member.type else 'undefined'+str(member.size))
```

Comment (mahaloz) [reply binsync/decompilers/ghidra/controller.py:149]: (this code should replace the entire if-stmt tree)

File: binsync/decompilers/ghidra/controller.py [hunk 4]
```diff
@@ -130,8 +130,40 @@ def fill_function_header(self, func_addr, user=None, artifact=None, decompilatio
...
+                        ghidra_struct.replaceAtOffset(offset, gtype, member.size, member.name, "")
+                        break
+            if old_ghidra_struct:
+                data_manager.replaceDataType(old_ghidra_struct, ghidra_struct, True)
+            else:
+                data_manager.addDataType(ghidra_struct, handler.DEFAULT_HANDLER)
+        except Exception as ex:
+            print(f'Error filling struct {struct_name}: {ex}')
+            changes = False
+        finally:
```
Comment (mahaloz) [inline binsync/decompilers/ghidra/controller.py:164]: finally dies here because the transaction decorator above

**Review by mahaloz — CHANGES_REQUESTED** [review]

File: binsync/decompilers/ghidra/controller.py (hunk 2 above)
Comment (Flipout50) [reply binsync/decompilers/ghidra/controller.py:152]: That is for field name though, this is for the type definition, i'm pretty sure ghidra requires the typestr for undefined types to be undefined + field size

**Pull request merged**